### PR TITLE
Bidirectional expand into

### LINKED ISSUE #1036: ExpandInto doesn't properly resolve bidirectional traversals
Given the graph created by:
```
CREATE ({v: 'src'})-[:E {v: 'edge'}]->({v: 'dest'})
```
The bidirectional conditional traversal works as expected:
```
redis-cli GRAPH.QUERY G "MATCH (a)-[e]-(b) RETURN a.v, e.v, b.v"
1) 1) "a.v"
   2) "e.v"
   3) "b.v"
2) 1) 1) "src"
      2) "edge"
      3) "dest"
   2) 1) "dest"
      2) "edge"
      3) "src"
```
But the logical equivalent with an ExpandInto causes an error:
```
"MATCH (a), (b) WITH a, b MATCH (a)-[e]-(b) RETURN a.v, e.v, b.v"
1) 1) "a.v"
   2) "e.v"
   3) "b.v"
2) 1) 1) "src"
      2) "edge"
      3) "dest"
3) (error) Type mismatch: expected a map but was Unknown
```

Depending on how the referenced edge `e` is accessed (such as with `RETURN e`), this can also manifest as an assertion failure.

The issue here is that ExpandInto doesn't have logic equivalent to https://github.com/RedisGraph/RedisGraph/blob/master/src/execution_plan/ops/op_conditional_traverse.c#L152-L159 .

This can be solved within ExpandInto, but I think it would be preferable to make shared logic functions for CondTraverse and ExpandInto to both use.

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Resolves #1036, introduces shared logic for populating referenced edges in CondTraverse and ExpandInto operations.

### CODE REVIEW

File: tests/flow/test_bidirectional_traversals.py [hunk 1]
```diff
@@ -310,3 +310,33 @@ def test11_bidirectional_multiple_edge_type(self):
...
+        query = """MATCH (a), (b) WITH a, b MATCH (a)-[e:E]-(b) RETURN ID(e), a.val, b.val ORDER BY ID(e), a.val, b.val"""
+        actual_result = acyclic_graph.query(query)
+        expected_result = [[0, 'v1', 'v2'],
+                           [0, 'v2', 'v1'],
+                           [1, 'v2', 'v3'],
+                           [1, 'v3', 'v2']]
+        self.env.assertEquals(actual_result.result_set, expected_result)
+
+        # Verify result against the equivalent conditional traversal.
+        query = """MATCH (a)-[e:E]-(b) RETURN ID(e), a.val, b.val ORDER BY ID(e), a.val, b.val"""
```
Comment (DvirDukhan) [inline tests/flow/test_bidirectional_traversals.py:340]: set `id` as edge property as we can't rely on the order of the `redisgraph-py` flushes the data

File: src/execution_plan/ops/shared/traverse_functions.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#pragma once
+
+#include "../../execution_plan.h"
+#include "../../../arithmetic/algebraic_expression.h"
+
+typedef struct {
```
Comment (DvirDukhan) [inline src/execution_plan/ops/shared/traverse_functions.h:12]: Please add a comment where this struct is used

File: src/execution_plan/ops/shared/traverse_functions.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+} EdgeTraverseData;
+
+// Initialize an EdgeTraverseData struct to populate edges appropriately for traversal operations.
+void Traverse_NewEdgeData(EdgeTraverseData *edge_data, AlgebraicExpression *ae,
+						  QGEdge *e, int idx);
+
+// Collect all appropriate edges between the given endpoints.
+void Traverse_CollectEdges(EdgeTraverseData *edge_data, NodeID src, NodeID dest);
+
+// If a matching edge is available, pop it and add it to the Record.
```
Comment (DvirDukhan) [inline src/execution_plan/ops/shared/traverse_functions.h:26]: `// Removes an available matching edge from edges array and sets it in the Record.`

File: tests/tck/features/MatchingSelfRelationships.feature [hunk 4]
```diff
@@ -63,6 +63,7 @@ Feature: MatchingSelfRelationships
       | 1        |
     And no side effects
 
+@skip
```
Comment (DvirDukhan) [inline tests/tck/features/MatchingSelfRelationships.feature:66]: did it pass by accident?

File: tests/tck/features/MatchingSelfRelationships.feature [hunk 5]
```diff
@@ -192,6 +193,7 @@ Feature: MatchingSelfRelationships
...
```
Comment (DvirDukhan) [inline tests/tck/features/MatchingSelfRelationships.feature:196]: did it pass by accident?

File: src/execution_plan/ops/shared/traverse_functions.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+typedef struct {
+	int *edgeRelationTypes;     // The relation type IDs that should be collected.
+	Edge *edges;                // Flexible array of all matching edges for the current endpoints.
+	int edgeIdx;                // The Record index for the referenced edge.
+	GRAPH_EDGE_DIR direction;   // The direction of the referenced edge being traversed.
+} EdgeTraverseData;
+
+// Initialize an EdgeTraverseData struct to populate edges appropriately for traversal operations.
+void Traverse_NewEdgeData(EdgeTraverseData *edge_data, AlgebraicExpression *ae,
```
Comment (DvirDukhan) [inline src/execution_plan/ops/shared/traverse_functions.h:20]: I think that returning an allocated struct is the way to go here.

File: src/execution_plan/ops/op_expand_into.c [hunk 7]
```diff
@@ -264,23 +210,15 @@ static void ExpandIntoFree(OpBase *ctx) {
...
-		array_free(op->edgeRelationTypes);
-		op->edgeRelationTypes = NULL;
-	}
-
 	if(op->ae) {
 		AlgebraicExpression_Free(op->ae);
 		op->ae = NULL;
 	}
 
+	if(op->setEdge) Traverse_FreeEdgeData(&op->edge_data);
```
Comment (DvirDukhan) [inline src/execution_plan/ops/op_expand_into.c:218]: I think that it is better to have a pointer to an allocated struct.
Call the free function regardless and remove line 189 as it is breaking abstraction

File: src/execution_plan/ops/op_conditional_traverse.h [hunk 8]
```diff
@@ -16,21 +17,17 @@ typedef struct {
...
-	int srcNodeIdx;             // Index into record.
-	int destNodeIdx;            // Index into record.
-	int *edgeRelationTypes;     // One or more relation types.
-	int edgeRelationCount;      // length of edgeRelationTypes.
 	GrB_Matrix F;               // Filter matrix.
 	GrB_Matrix M;               // Algebraic expression result.
-	bool setEdge;               // Edge needs to be set.
-	Edge *edges;                // Discovered edges.
+	bool setEdge;               // Edge needs to be set in the Record.
+	EdgeTraverseData edge_data; // Edge collection data if the edge needs to be set.
```
Comment (DvirDukhan) [inline src/execution_plan/ops/op_conditional_traverse.h:23]: I think that it is better to have a pointer to an allocated struct.

File: src/execution_plan/ops/shared/traverse_functions.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "../../execution_plan.h"
+#include "../../../arithmetic/algebraic_expression.h"
+
+typedef struct {
+	int *edgeRelationTypes;     // The relation type IDs that should be collected.
+	Edge *edges;                // Flexible array of all matching edges for the current endpoints.
+	int edgeIdx;                // The Record index for the referenced edge.
+	GRAPH_EDGE_DIR direction;   // The direction of the referenced edge being traversed.
+} EdgeTraverseData;
```
Comment (DvirDukhan) [inline src/execution_plan/ops/shared/traverse_functions.h:17]: Maybe `TraverseEdgeContext` is more suitable?

File: src/execution_plan/ops/op_conditional_traverse.c [hunk 10]
```diff
@@ -267,23 +207,15 @@ static void CondTraverseFree(OpBase *ctx) {
...
 	if(op->ae) {
 		AlgebraicExpression_Free(op->ae);
 		op->ae = NULL;
 	}
 
-	if(op->edgeRelationTypes) {
-		array_free(op->edgeRelationTypes);
-		op->edgeRelationTypes = NULL;
-	}
+	if(op->setEdge) Traverse_FreeEdgeData(&op->edge_data);
```
Comment (DvirDukhan) [inline src/execution_plan/ops/op_conditional_traverse.c:215]: please call Free regardless, remove the `array_clear` from line 177 as it is breakinging abstraction
use a pointer to an allocated struct

File: tests/tck/features/MatchingSelfRelationships.feature (hunk 4 above)
Comment (jeffreylovitz) [reply tests/tck/features/MatchingSelfRelationships.feature:66]: Yes. Bidirectional traversals typically return each edge twice (incoming and outgoing), but self-loops in Cypher still should only return once - I'm guessing this is a consequence of the edge uniqueness constraint we don't currently adhere to. I conferred with Roi and we agreed that these should be skips for the moment.

File: src/execution_plan/ops/shared/traverse_functions.h (hunk 9 above)
Comment (jeffreylovitz) [reply src/execution_plan/ops/shared/traverse_functions.h:17]: :+1: That's better, thanks!

File: src/execution_plan/ops/shared/traverse_functions.c [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+void Traverse_ResetEdgeCtx(EdgeTraverseCtx *edge_ctx) {
+	array_clear(edge_ctx->edges);
+}
+
+void Traverse_FreeEdgeCtx(EdgeTraverseCtx *edge_ctx) {
+	if(!edge_ctx) return;
+
+	array_free(edge_ctx->edges);
```
Comment (DvirDukhan) [inline src/execution_plan/ops/shared/traverse_functions.c:93]: maybe I missed it. Do we gurentee for the array to be empty? Do we need to free the edges?

Comment (jeffreylovitz) [reply src/execution_plan/ops/shared/traverse_functions.c:93]: The Edges themselves aren't heap allocations, they're contiguous structs in this array: https://github.com/RedisGraph/RedisGraph/blob/master/src/graph/graph.c#L204-L206

File: src/execution_plan/ops/op_expand_into.h [hunk 12]
```diff
@@ -7,6 +7,7 @@
 #pragma once
 
 #include "op.h"
+#include "shared/traverse_functions.h"
 #include "../execution_plan.h"
 #include "../../graph/graph.h"
 #include "../../graph/entities/edge.h"
```
Comment (swilly22) [inline src/execution_plan/ops/op_expand_into.h:13]: Please go over this include list, see if all are necessary.

File: src/execution_plan/ops/op_expand_into.c [hunk 13]
```diff
@@ -97,30 +69,27 @@ OpBase *NewExpandIntoOp(const ExecutionPlan *plan, Graph *g, AlgebraicExpression
...
 	op->ae = ae;
 	op->r = NULL;
-	op->edges = NULL;
 	op->F = GrB_NULL;
 	op->M = GrB_NULL;
 	op->recordCount = 0;
-	op->edgeRelationTypes = NULL;
-	op->recordsCap = 0;
 	op->records = NULL;
+	op->recordsCap = 0;
```
Comment (swilly22) [inline src/execution_plan/ops/op_expand_into.c:76]: It feels a waste to use `rm_calloc` but then set explicit NULLs,
if we still want to do this, we should set `edge_ctx` to NULL at this point.

File: src/execution_plan/ops/op_conditional_traverse.c [hunk 14]
```diff
@@ -18,62 +18,12 @@ static OpResult CondTraverseReset(OpBase *opBase);
...
-		// If we're traversing incoming edges, swap the source and destination.
-		__CondTraverse_CollectEdges(op, op->destNodeIdx, op->srcNodeIdx);
-		return;
-	case GRAPH_EDGE_DIR_BOTH:
-		// If we're traversing in both directions, collect edges in both directions.
-		__CondTraverse_CollectEdges(op, op->srcNodeIdx, op->destNodeIdx);
-		__CondTraverse_CollectEdges(op, op->destNodeIdx, op->srcNodeIdx);
-		return;
-	}
+static inline int CondTraverseToString(const OpBase *ctx, char *buf, uint buf_len) {
```
Comment (swilly22) [inline src/execution_plan/ops/op_conditional_traverse.c:21]: Although this function is basically a proxy function, it doesn't call too often for it to be declared as `inline`

File: src/execution_plan/ops/op_conditional_traverse.c [hunk 15]
```diff
@@ -117,22 +67,15 @@ void _traverse(CondTraverse *op) {
...
 	op->ae = ae;
 	op->r = NULL;
 	op->iter = NULL;
-	op->edges = NULL;
 	op->F = GrB_NULL;
 	op->M = GrB_NULL;
-	op->recordsLen = 0;
-	op->direction = GRAPH_EDGE_DIR_OUTGOING;
-	op->edgeRelationTypes = NULL;
+	op->recordCount = 0;
```
Comment (swilly22) [inline src/execution_plan/ops/op_conditional_traverse.c:78]: See comment on `NewExpandIntoOp`

**Pull request merged**